azure/converters/spotinstances.go

```diff
@@ -40,8 +40,12 @@ func GetSpotVMOptions(spotVMOptions *infrav1.SpotVMOptions, diffDiskSettings *in
 			MaxPrice: &maxPrice,
 		}
 	}
+
+	// Set the spot vm eviction policy if provided, otherwise default to deallocate for regular disks and delete for ephemeral disks
 	evictionPolicy := compute.VirtualMachineEvictionPolicyTypesDeallocate
```

Comment:
rethinking this, it would be better to explicitly default the field in the defaulting webhook so that the value is persisted in the Spec, even if the user didn't explicitly set it. That way, if we decide to change the default later, we won't break existing users.

Reply:
Webhook defaulting logic for AzureMachines: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_default.go
Webhook defaulting logic for AzureMachinePools: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_default.go

More info on webhooks: https://cluster-api.sigs.k8s.io/developer/providers/webhooks.html

Reply:
If we define the default in the webhook, does that mean we won't have to do any checks inside the spotinstances converter, and just return the eviction policy from the spotVMOptions parameter directly? I'm assuming that by the time the converter is called when building the VM from spec, the webhook will set the default spotVMOptions if it isn't defined in the spec.

Reply:
I'm also wondering if we should still default to delete if an ephemeral disk is being used. If the spec is already defined by the webhook by default, I'm not sure how to check if that's the case before setting it to delete for an ephemeral disk.

Reply:
> If we define the default in the webhook, does that mean we won't have to do any checks inside the spotinstances converter, and just return the eviction policy from the spotVMOptions parameter directly

correct

> I'm also wondering if we should still default to delete if an ephemeral disk is being used

you can make the webhook default a "smart" default where it defaults to delete if ephemeral disk is being used

Reply:
@CecileRobertMichon I was able to make changes to the webhooks to set the default spot eviction policy. However, many of the unit tests started failing because I don't think the defaulting webhook is run for them. Is there a way to run the webhook in a unit test environment?

Reply:
see my comment below https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2689/files#r990558486